Sources/GRPC/ClientConnection.swift

```diff
@@ -53,139 +53,239 @@ import NIOTLS
 /// delegate associated with this connection (see `DelegatingErrorHandler`).
 ///
 /// See `BaseClientCall` for a description of the remainder of the client pipeline.
-open class ClientConnection {
-  /// Makes and configures a `ClientBootstrap` using the provided configuration.
-  ///
-  /// Enables `SO_REUSEADDR` and `TCP_NODELAY` and configures the `channelInitializer` to use the
-  /// handlers detailed in the documentation for `ClientConnection`.
-  ///
-  /// - Parameter configuration: The configuration to prepare the bootstrap with.
-  public class func makeBootstrap(configuration: Configuration) -> ClientBootstrapProtocol {
-    let bootstrap = GRPCNIO.makeClientBootstrap(group: configuration.eventLoopGroup)
-      // Enable SO_REUSEADDR and TCP_NODELAY.
-      .channelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
-      .channelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
-      .channelInitializer { channel in
-        let tlsConfigured = configuration.tlsConfiguration.map { tlsConfiguration in
-          channel.configureTLS(tlsConfiguration, errorDelegate: configuration.errorDelegate)
+public class ClientConnection {
+  /// The configuration this connection was created using.
+  internal let configuration: ClientConnection.Configuration
+
+  /// The channel which will handle gRPC calls.
+  internal var channel: EventLoopFuture<Channel>
+
+  /// HTTP multiplexer from the `channel` handling gRPC calls.
+  internal var multiplexer: EventLoopFuture<HTTP2StreamMultiplexer>
+
+  /// A monitor for the connectivity state.
+  public let connectivity: ConnectivityStateMonitor
+
+  /// Creates a new connection from the given configuration.
+  public init(configuration: ClientConnection.Configuration) {
+    let monitor = ConnectivityStateMonitor(delegate: configuration.connectivityStateDelegate)
+    let channel = ClientConnection.makeChannel(
+      configuration: configuration,
+      connectivityMonitor: monitor
+    )
+
+    self.channel = channel
+    self.multiplexer = channel.flatMap {
+      $0.pipeline.handler(type: HTTP2StreamMultiplexer.self)
+    }
+    self.connectivity = monitor
+    self.configuration = configuration
+
+    self.channel.whenSuccess { _ in
+      self.connectivity.state = .ready
+    }
+    self.replaceChannelAndMultiplexerOnClose(channel: channel)
+  }
+
+  /// Registers a callback on the `closeFuture` of the given channel to replace this class's
+  /// channel and multiplexer.
+  private func replaceChannelAndMultiplexerOnClose(channel: EventLoopFuture<Channel>) {
+    channel.always { result in
+      // If we failed to get a channel then we've exhausted our backoff; we should `.shutdown`.
+      if case .failure = result {
+        self.connectivity.state = .shutdown
+      }
+    }.flatMap {
+      $0.closeFuture
+    }.whenComplete { _ in
+      // `.shutdown` is terminal so don't attempt a reconnection.
```

Comment:
This looks very similar to `replaceChannelAndMultiplexer()`, any chance to merge the two? (If they are too different, just let me know.)

Also, when is `replaceChannelAndMultiplexer()` even called?

Reply:
Agh, sorry, `replaceChannelAndMultiplexer` was meant to be removed.